pkg/util/ssh.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+
+	"github.com/golang/glog"
+	"golang.org/x/crypto/ssh"
+)
+
+// TODO: Unit tests for this code, we can spin up a test SSH server with instructions here:
+// https://godoc.org/golang.org/x/crypto/ssh#ServerConn
+type SSHTunnel struct {
+	Config     *ssh.ClientConfig
+	Host       string
+	SSHPort    int
+	LocalPort  int
+	RemoteHost string
+	RemotePort int
+	running    bool
+	sock       net.Listener
+	client     *ssh.Client
+}
+
+func (s *SSHTunnel) copyBytes(out io.Writer, in io.Reader) {
+	if _, err := io.Copy(out, in); err != nil {
+		glog.Errorf("Error in SSH tunnel: %v", err)
+	}
+}
+
+func NewSSHTunnel(user, keyfile, host, remoteHost string, localPort, remotePort int) (*SSHTunnel, error) {
+	signer, err := MakePrivateKeySigner(keyfile)
+	if err != nil {
+		return nil, err
+	}
+	config := ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{ssh.PublicKeys(signer)},
+	}
+	return &SSHTunnel{
+		Config:     &config,
+		Host:       host,
+		SSHPort:    22,
+		LocalPort:  localPort,
+		RemotePort: remotePort,
+		RemoteHost: remoteHost,
+	}, nil
+}
+
+func (s *SSHTunnel) Open() error {
+	var err error
+	s.client, err = ssh.Dial("tcp", fmt.Sprintf("%s:%d", s.Host, s.SSHPort), s.Config)
```

Comment:
Should these be net.JoinHostPort()?

Won't matter in our case, but might be nice if any of those IPv6ers want to use this util package.